brave/src/main/java/brave/propagation/ExtraFieldPropagation.java

```diff
@@ -97,21 +96,73 @@
  * }</pre>
  */
 public final class ExtraFieldPropagation<K> implements Propagation<K> {
+  static final FieldCustomizer NOOP_FIELD_CUSTOMIZER = new FieldCustomizer() {
+    @Override public String update(String key, String value) {
+      return value;
+    }
+  };
+  static final Customizer[] NO_CONTEXT_CUSTOMIZERS = new Customizer[0];
+  static final FieldCustomizer[] NO_FIELD_CUSTOMIZERS = new FieldCustomizer[0];
+
+  public interface FieldCustomizer {
+    /**
+     * Returns the same value, an updated one, or null to drop the field.
+     *
+     * @see ExtraFieldPropagation#set(String, String)
+     */
+    @Nullable String update(String fieldName, @Nullable String value);
+  }
+
+  /**
+   * Methods here are called once per extract or inject, which allows the returned customizer to
+   * handle multiple fields as needed.
+   */
+  public static abstract class Customizer {
+    public FieldCustomizer extractCustomizer(TraceContextOrSamplingFlags.Builder builder) {
+      return NOOP_FIELD_CUSTOMIZER;
+    }
+
+    public FieldCustomizer injectCustomizer(TraceContext context) {
+      return NOOP_FIELD_CUSTOMIZER;
+    }
+  }
+
+  static final class RedactOnInject extends Customizer implements FieldCustomizer {
+    final LinkedHashSet<String> redactedFields;
+
+    RedactOnInject(LinkedHashSet<String> redactedFields) {
+      this.redactedFields = redactedFields;
+    }
+
+    @Override public FieldCustomizer injectCustomizer(TraceContext context) {
+      return this;
+    }
+
+    @Override public String update(String fieldName, String value) {
+      if (redactedFields.contains(fieldName)) return null;
+      return value;
+    }
+
+    @Override public String toString() {
+      return "RedactOnInject{" + redactedFields + "}";
+    }
+  }
+
   /** Wraps an underlying propagation implementation, pushing one or more fields */
   public static Factory newFactory(Propagation.Factory delegate, String... fieldNames) {
     if (delegate == null) throw new NullPointerException("delegate == null");
     if (fieldNames == null) throw new NullPointerException("fieldNames == null");
     String[] validated = ensureLowerCase(new LinkedHashSet<>(Arrays.asList(fieldNames)));
-    return new Factory(delegate, validated, validated, new BitSet());
```

Comment:
bitset is likely faster to do redaction, but I'm doing something different here. redaction via deleting fields is a very small edge case requested I think only by one person. In order to add the ability to customize propagation we need to make room for more complexity. This trades slightly better performance given by bit set in an edge case, for a more general feature that can be used for things besides redaction.